File: src/test/java/me/piclane/logview/procedure/text/LineTest.java
package me.piclane.logview.procedure.text;

import com.google.common.jimfs.Configuration;
import com.google.common.jimfs.Jimfs;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystem;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class LineTest {
    /** Test FileSystem */
    private FileSystem testFs;

    /** Test file path */
    private Path testPath;

    /** lines in test file */
    private List<String> expectLines;

    @Before
    public void before() throws IOException {
        StringBuilder buf = new StringBuilder();
        for(int i=0; i<10; i++) {
            for(int j=0; j<10; j++) {
                buf.append(i);
            }
            buf.append("\n");
        }
        String expectString = buf.toString();

        this.testFs = Jimfs.newFileSystem(Configuration.unix());
        this.testPath = testFs.getPath("/test.txt");
        try(BufferedWriter out = Files.newBufferedWriter(testPath)) {
            out.append(expectString);
        }
        expectLines = Files.readAllLines(testPath);
    }

    @After
    public void after() throws IOException {
        testFs.close();
    }

    @Test
    public void testReadLine() throws IOException {
        try(BufferedByteReader reader = new BufferedByteReader(testPath)) {
            List<String> actualLines = new ArrayList<>();
            AtomicInteger lineNum = new AtomicInteger();
            Line.readLine(reader, StandardCharsets.UTF_8, line -> {
                actualLines.add(line.toString());
                assertThat(line.pos, is(equalTo(lineNum.getAndIncrement() * 11L)));
                assertThat(line.len, is(equalTo(11L)));
                return true;
            });
            assertThat(actualLines, is(contains(expectLines.toArray())));
        }
    }

    @Test
    public void testSkipLine() throws IOException {
        try(BufferedByteReader reader = new BufferedByteReader(testPath)) {
            for(int lineNum = 1; lineNum<=10; lineNum++) {
                long pos = Line.skipLine(reader);
                assertThat(pos, is(equalTo(lineNum * 11L)));
            }
        }
    }
}

File: src/test/kotlin/com/xxuz/piclane/logview/procedure/text/LineTest.kt
package com.xxuz.piclane.logview.procedure.text

import com.google.common.jimfs.Configuration
import com.google.common.jimfs.Jimfs
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.FileSystem
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class LineTest {
    /** Test FileSystem  */
    private lateinit var testFs: FileSystem

    /** Test file path  */
    private lateinit var testPath: Path

    /** lines in test file  */
    private lateinit var expectLines: List<String>
    
    @BeforeEach
    fun before() {
        val buf = StringBuilder()
        for (i in 0..9) {
            for (j in 0..9) {
                buf.append(i)
            }
            buf.append("\n")
        }
        val expectString = buf.toString()
        testFs = Jimfs.newFileSystem(Configuration.unix())
        testPath = testFs.getPath("/test.txt")
        Files.newBufferedWriter(testPath).use { out -> out.append(expectString) }
        expectLines = Files.readAllLines(testPath)
    }

    @AfterEach
    fun after() {
        testFs.close()
    }

    @Test
    fun testReadLine() {
        BufferedByteReader(testPath).use { reader ->
            val actualLines: MutableList<String> = ArrayList()
            val lineNum = AtomicInteger()
            Line.readLine(reader, StandardCharsets.UTF_8) { line ->
                actualLines.add(line.toString())
                MatcherAssert.assertThat(line.pos, Matchers.`is`(Matchers.equalTo(lineNum.getAndIncrement() * 11L)))
                MatcherAssert.assertThat(line.len, Matchers.`is`(Matchers.equalTo(11L)))
                true
            }
            MatcherAssert.assertThat<List<String>>(actualLines, Matchers.`is`(Matchers.contains<Any>(*expectLines.toTypedArray())))
        }
    }

    @Test
    fun testSkipLine() {
        BufferedByteReader(testPath).use { reader ->
            for (lineNum in 1..10) {
                val pos = Line.skipLine(reader)
                MatcherAssert.assertThat(pos, Matchers.`is`(Matchers.equalTo(lineNum * 11L)))
            }
        }
    }
}
